File: 5.最长回文子串.cpp
/*
 * @lc app=leetcode.cn id=5 lang=cpp
 *
 * [5] 最长回文子串
 */
#include<string>
#include<algorithm>
#include<vector>
using namespace std;
// @lc code=start
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size(), maxLength = 1, begin = 0;
        vector<vector<bool>> dp(n, vector<bool>(n, false));
        for (int i = 0; i < n; i++)
            dp[i][i] = true;
        for (int l = 2; l <= s.length(); l++) {
            for (int i = 0; i < s.length(); i++) {
                int j = i + l - 1;
                if (j >= s.length())
                    break;
                if (s[i] == s[j]) {
                    if (l <= 2) 
                        dp[i][j] = true;
                    else
                        dp[i][j] = dp[i + 1][j - 1];
                }
                if (dp[i][j] && l > maxLength) {
                    maxLength = l;
                    begin = i;
                }
            }
        }
        return s.substr(begin, maxLength);
    }
};
// @lc code=end

File: 5.最长回文子串.java
/*
 * @lc app=leetcode.cn id=5 lang=java
 *
 * [5] 最长回文子串
 */

// @lc code=start
class Solution {
    public String longestPalindrome(String s) {
        String res = "";
        for (int i = 0; i < s.length(); i++) {
            String sing = palindrome(s, i, i);
            res = res.length() > sing.length() ? res : sing;
            String doub = palindrome(s, i, i + 1);
            res = res.length() > doub.length() ? res : doub;
        }
        return res;
    }

    String palindrome(String s, int l, int r) {
        while (l >= 0 && r < s.length() && s.charAt(r) == s.charAt(l)) {
            l--;
            r++;
        }
        return s.substring(l + 1, r);
    }
}
// @lc code=end

